libr/bin/p/bin_pef.c

```diff
@@ -0,0 +1,769 @@
+		ret->machine = strdup("Macintosh");
+	}
+	ret->bits = 32;
+	ret->has_pi = true;
+	ret->big_endian = true;
+	ret->rclass = strdup("cfm");
+	ret->has_va = true;
+	return ret;
+}
+
+static RList *fields(RBinFile *bf) {
+	RList *ret = r_list_newf((RListFree)free);
+	#define ROW(nam, siz, val, fmt, cmt) \
+		r_list_append (ret, r_bin_field_new (addr, addr, val, siz, nam, cmt, fmt, false));
+	ut64 addr = 0;
+	ROW("magic", 8, r_buf_read_be64_at(bf->buf, addr), "x", NULL); addr += 8;
+	ROW("architecture", 4, r_buf_read_be32_at(bf->buf, addr), "x", NULL); addr += 4;
+	ROW("formatVersion", 4, r_buf_read_be32_at(bf->buf, addr), "x", NULL); addr += 4;
+	ROW("dateTimeStamp", 4, r_buf_read_be32_at(bf->buf, addr), "x", NULL); addr += 4;
+	ROW("oldDefVersion", 4, r_buf_read_be32_at(bf->buf, addr), "x", NULL); addr += 4;
+	ROW("oldImpVersion", 4, r_buf_read_be32_at(bf->buf, addr), "x", NULL); addr += 4;
+	ROW("currentVersion", 4, r_buf_read_be32_at(bf->buf, addr), "x", NULL); addr += 4;
+	ROW("sectionCount", 2, r_buf_read_be16_at(bf->buf, addr), "x", NULL); addr += 2;
+	ROW("instSectionCount", 2, r_buf_read_be16_at(bf->buf, addr), "x", NULL); addr += 2;
+	ROW("reserved", 2, r_buf_read_be16_at(bf->buf, addr), "x", NULL); addr += 2;
+	return ret;
+}
+
+// There is a "preferred base address" in PEF section headers but it is never used.
+static ut64 baddr(RBinFile *bf) {
+	return 0;
+}
+
+static ut64 size(RBinFile *bf) {
+	RBinPEFObj *pef = bf->bo->bin_obj;
+	ut64 s = 0;
+	for (int i=0; i<pef->nsec; i++) {
+		ut64 e = pef->sec[i].offset + pef->sec[i].lenDisk;
+		if (s < e) s = e;
+	}
+	return s;
+}
+
+static RBinAddr *binsym(RBinFile *bf, int sym) {
+	RBinPEFObj *pef = bf->bo->bin_obj;
+
+	int n;
+	if (sym == R_BIN_SYM_ENTRY || sym == R_BIN_SYM_MAIN) {
+		n = 0;
+	} else if (sym == R_BIN_SYM_INIT) {
+		n = 1;
+	} else if (sym == R_BIN_SYM_FINI) {
+		n = 2;
+	} else {
+		return NULL;
+	}
+
+	ut32 sec = r_buf_read_be32_at(bf->buf, pef->ldrsec + n*8);
+	ut32 offset = r_buf_read_be32_at(bf->buf, pef->ldrsec + n*8+4);
+	if (sec < pef->nsec && offset < pef->sec[sec].addr+pef->sec[sec].lenTotal) {
+		RBinAddr *ptr = R_NEW0(RBinAddr);
+		ptr->vaddr = pef->sec[sec].addr + offset;
+		return ptr;
+	} else {
+		return NULL;
+	}
+}
+
+static RList *sections(RBinFile *bf) {
+	RBinPEFObj *pef = bf->bo->bin_obj;
+	RList *ret = r_list_newf((RListFree)r_bin_section_free);
+
+	for (ut16 i=0; i<pef->nsec; i++) {
+		PEFSection *sec = &pef->sec[i];
+		RBinSection *ptr = R_NEW0(RBinSection);
+		ptr->is_segment = true;
+
+		ptr->paddr = sec->offset;
+		ptr->size = sec->lenDisk;
+		ptr->vaddr = sec->addr;
+		ptr->vsize = sec->lenTotal;
+
+		ptr->type = strdup(
+			sec->kind == 0 ? "TEXT" :
+			sec->kind == 1 ? "DATA" :
+			sec->kind == 2 ? "PIDATA" :
+			sec->kind == 3 ? "RODATA" :
+			sec->kind == 4 ? "LOADER" :
+			"");
+		ptr->name = strdup(ptr->type);
+		ptr->arch = strdup("ppc");
+		ptr->is_data = sec->kind != 0;
+		ptr->add = sec->kind != 4;
+		ptr->perm =
```

Comment:
maybe better to have a switch statement to avoid multiple dereferences, or cache the value in a const variable first.. but i think a switch will be more readable